### PR TITLE
Post/page: update inserted image url if edited

### LINKED ISSUE #9036: Image Editor: Update image url if post/page is being edited
In the case of:
1. We are editing a post or page in the Calypso Editor
2. We insert an image using the media modal
3. We edit the image within the Calypso Editor

We should update the image reference to match the new edit for that open post or page. In this task we do not need to update other references in other posts/pages. 
(Say the image is inserted in both post 4 and 5. However we are editing post 5, when we update the image. I'd expect only the reference in post 5 to be updated.)

This matches the current behavior of wp-admin.

### PR DESCRIPTION
**Pull request by lamosty**
Fixes #9036. After editing an inserted image in a post/page, let's update its url in the post/page as it is changed on the backend.

## Testing instructions

1. Open a post/page and insert an image;
2. Edit the image in image editor and click "Done";
3. Verify that the inserted image has correct (updated) url

## Todo

- [x] After editing an image, update it in post/page as a transient one (use the edited image from client if possible). Update the image again once it has been successfully uploaded to a server.

/cc @gwwar @aduth @timmyc

### CODE REVIEW

**Comment by gwwar** [comment]
Pretty cool. Very close, but it looks like we can some interesting sizes on an edit:
https://cldup.com/7XnlHOAjKF.gif

**Comment by gwwar** [comment]
👍  Very nice, looks like the image is updated to an expected size. Let's wait for some feedback from @aduth or @timmyc before 🚢  in case we're missing a few more cases to test

File: client/components/tinymce/plugins/media/plugin.jsx [hunk 1]
```diff
@@ -192,8 +192,8 @@ function mediaButton( editor ) {
 
 				// When merging, allow any updated field to be used if it doesn't
 				// already exist in the current markup, but otherwise only force
-				// update ID and URL attributes to their new values
-				const merged = assign( {}, media, current.media, pick( media, 'ID', 'URL' ), {
+				// update ID, URL, width and height attributes to their new values
+				const merged = assign( {}, media, current.media, pick( media, 'ID', 'URL', 'width', 'height' ), {
```
Comment (gwwar) [inline client/components/tinymce/plugins/media/plugin.jsx]: Actually I found another case:
1. Throttle your connection to something slow
2. Drag and drop a large image directly into the post content
3. Resize to something much smaller
4. Set network connection back to no throttle
5. We update the image back to full size

Comment (retrofox) [inline client/components/tinymce/plugins/media/plugin.jsx]: Rebased with master since I've deployed `restore image` feature a few minutes ago. The restored image isn't updated into the post content.

File: client/components/tinymce/plugins/media/plugin.jsx [hunk 2]
```diff
@@ -135,6 +136,8 @@ function mediaButton( editor ) {
 		return { isLoaded, onLoad };
 	} )();
 
+	const mediaHasChanged = ( media ) => withoutHttp( media.guid ) !== withoutHttp( media.URL );
```
Comment (lamosty) [inline client/components/tinymce/plugins/media/plugin.jsx]: Should we move this to a separate util file? If yes, which one/where?

Comment (retrofox) [inline client/components/tinymce/plugins/media/plugin.jsx]: Once the media is edited for the first time it always will return `true`.

I've added a console and just in some `initial` states its value is `true`.

image

Also, I've edited the code returning always `true` and I don't see any change in terms of the functionality.

image

And btw, it works!. So I wonder if we should just remove this helper function.

Comment (lamosty) [inline client/components/tinymce/plugins/media/plugin.jsx]: Haha, nice. Nono, we need it. Yeah, for the line 184 (`	} else if ( ! mediaHasChanged( media ) ) {`), it doesn't do much (as you've pointed out), but for the media width/height, it's still important.

Try this:
1. Drag and drop new media of bigger size (>2mb) or throttle your network connection;
2. Resize the transient media which was inserted into the post editor;
3. Unthrottle your network and watch the media get uploaded;
4. Its size in post editor shouldn't change;
5. Now, edit the image in image editor to a very small size (smaller than the size of the inserted image in post editor);
6. The size of the inserted image should change to reflect the size change when you edited the image;
7. On the other hand, if the edited image has larger size than the inserted image, don't change the size of the inserted image.

File: client/components/tinymce/plugins/media/plugin.jsx [hunk 3]
```diff
@@ -190,12 +188,33 @@ function mediaButton( editor ) {
 					return;
 				}
 
+				let useMediaWidthHeight = false;
```
Comment (omarjackman) [inline client/components/tinymce/plugins/media/plugin.jsx]: maybe also change it to `useMediaSize`

File: client/components/tinymce/plugins/media/plugin.jsx [hunk 4]
```diff
@@ -190,12 +188,33 @@ function mediaButton( editor ) {
...
+				// update ID, URL, width and height attributes to their new values
+				const merged = assign(
+					{},
+					media,
+					current.media,
+					pick(
+						media,
+						'ID',
+						'URL',
+						useMediaWidthHeight ? 'width' : '',
```
Comment (omarjackman) [inline client/components/tinymce/plugins/media/plugin.jsx]: also maybe `useMediaWidthHeight && 'width'` instead of the ternary op

Comment (lamosty) [inline client/components/tinymce/plugins/media/plugin.jsx]: I'd like to explicitly pass empty string `''` if `useMediaWidthHeight` is false instead of `false`.

**Comment by gwwar** [comment]
Whoops, I also forgot about caption sizing:
screen shot 2016-11-03 at 1 33 30 pm

1. Insert an image with a caption
2. Edit it to be smaller
3. Caption is original size

File: client/components/tinymce/plugins/media/plugin.jsx [hunk 5]
```diff
@@ -135,6 +135,69 @@ function mediaButton( editor ) {
 		return { isLoaded, onLoad };
 	} )();
 
+	function parseShortcode( content ) {
```
Comment (lamosty) [inline client/components/tinymce/plugins/media/plugin.jsx]: I copied this function from `tinymce/plugins/wpeditimage/plugin`. We should move it to a shared helper file. However, it depends on `tinymce` so might be large.

File: client/components/tinymce/plugins/media/plugin.jsx [hunk 6]
```diff
@@ -145,10 +208,19 @@ function mediaButton( editor ) {
...
+		let numOfImagesToUpdate = 0;
+
 		isVisualEditMode = ! editor.isHidden();
 
 		if ( isVisualEditMode ) {
-			images = editor.dom.select( 'img' );
+			if ( dirtyImage ) {
+				images = editor.dom.select( `img.wp-image-${ dirtyImage.ID }` );
+
+				numOfImagesToUpdate = images.length;
```
Comment (lamosty) [inline client/components/tinymce/plugins/media/plugin.jsx]: We need to remember how many dirty images there are because after we update all of them, we need to set the image media `.isDirty` flag to `false` so the images won't be updated every time another image is going to be updated.

File: client/components/tinymce/plugins/media/plugin.jsx [hunk 7]
```diff
@@ -168,14 +240,28 @@ function mediaButton( editor ) {
...
 			}
 
+			const media = MediaStore.get( selectedSite.ID, current.media.ID );
+			let mediaHasCaption = false;
+			let captionNode = null;
+
+			if ( media.isDirty ) {
+				current.media.transient = true;
+
+				captionNode = editor.dom.getParent( img, '.mceTemp' );
```
Comment (lamosty) [inline client/components/tinymce/plugins/media/plugin.jsx]: The caption hack: if we want to keep captions sizing and alignment in sync with the updated images, we need to remove and add new caption shortcode on each image edit.

File: client/components/tinymce/plugins/media/plugin.jsx [hunk 8]
```diff
@@ -190,22 +276,50 @@ function mediaButton( editor ) {
...
+					}
+				);
 				const options = assign( {}, current.appearance, {
 					forceResize: ! media.transient && current.media.width && current.media.width !== media.width
 				} );
 
-				// Since we're explicitly targetting the image, don't allow the
-				// caption to be considered in the generated markup
-				delete merged.caption;
+				if ( ! mediaHasCaption ) {
```
Comment (lamosty) [inline client/components/tinymce/plugins/media/plugin.jsx]: Delete the caption only if we are not updating an image or if the updated image doesn't have a caption.

File: client/components/tinymce/plugins/media/plugin.jsx [hunk 9]
```diff
@@ -190,22 +276,50 @@ function mediaButton( editor ) {
...
+					// Since we're explicitly targetting the image, don't allow the
+					// caption to be considered in the generated markup
+					delete merged.caption;
+				}
 
 				// Use markup utility to generate replacement element
 				markup = MediaMarkup.get( merged, options );
```
Comment (lamosty) [inline client/components/tinymce/plugins/media/plugin.jsx]: If `merged` contains `caption` prop, the `img` tag will get wrapped in the caption shortcode with correct sizing.

File: client/components/tinymce/plugins/media/plugin.jsx [hunk 10]
```diff
@@ -223,7 +337,7 @@ function mediaButton( editor ) {
 
 			// To avoid an undesirable flicker after the image uploads but
 			// hasn't yet been loaded, we preload the image before rendering.
-			const imageUrl = MediaSerialization.deserialize( event.content ).media.URL;
```
Comment (lamosty) [inline client/components/tinymce/plugins/media/plugin.jsx]: This would no longer work if `event.content` contains also the `[caption]` shortcode.

File: client/components/tinymce/plugins/media/plugin.jsx [hunk 11]
```diff
@@ -234,32 +348,45 @@ function mediaButton( editor ) {
...
-			let imgString;
-			if ( img.outerHTML ) {
-				imgString = img.outerHTML;
+			let mediaString = '';
+
+			if ( mediaHasCaption ) {
+				if ( captionNode.outerHTML ) {
+					mediaString = captionNode.outerHTML;
+
+					captionNode.outerHTML = event.content;
```
Comment (lamosty) [inline client/components/tinymce/plugins/media/plugin.jsx]: This is how we update the post/page editor image.

Now, if the updated image also has a caption, let's target the caption wrapper and replace its `outerHTML` with the new markup. Otherwise, continue as before.

File: client/components/tinymce/plugins/media/plugin.jsx [hunk 12]
```diff
@@ -234,32 +348,45 @@ function mediaButton( editor ) {
...
 			// instance in the undo stack.
 			//
 			// See: https://github.com/tinymce/tinymce/blob/4.2.4/js/tinymce/classes/EditorUpload.js#L49-L53
 			editor.undoManager.data = editor.undoManager.data.map( function( level ) {
-				level.content = level.content.replace( imgString, event.content );
+				level.content = level.content.replace( mediaString, event.content );
 				return level;
 			} );
+
+			numOfImagesToUpdate -= 1;
```
Comment (lamosty) [inline client/components/tinymce/plugins/media/plugin.jsx]: Decrease the number of images that needs updating so we know when to set the media image `.isDirty` prop to `false`.

File: client/components/tinymce/plugins/media/plugin.jsx [hunk 13]
```diff
@@ -279,6 +406,10 @@ function mediaButton( editor ) {
 			// autosave take effect
 			editor.fire( 'change' );
 		}
+
+		if ( numOfImagesToUpdate === 0 ) {
+			MediaActions.edit(  selectedSite.ID, { ...dirtyImage, isDirty: false } );
```
Comment (lamosty) [inline client/components/tinymce/plugins/media/plugin.jsx]: If there are no more images to update, let's set the image `.isDirty` prop to false so the image is not updated each time other images are updated.

File: client/components/tinymce/plugins/media/plugin.jsx (hunk 5 above)
Comment (gwwar) [inline client/components/tinymce/plugins/media/plugin.jsx]: I'd be wary of copying this over. Is it possible to trigger a resize with an increment of 0? https://github.com/Automattic/wp-calypso/blob/master/client/components/tinymce/plugins/media/plugin.jsx#L387

Comment (gwwar) [inline client/components/tinymce/plugins/media/plugin.jsx]: Looking further, the selection won't be set, but it appears we're parsing via: `const parsed = MediaSerialization.deserialize( node );`

File: client/components/tinymce/plugins/media/plugin.jsx [hunk 14]
```diff
@@ -145,10 +208,19 @@ function mediaButton( editor ) {
 			return;
 		}
 
+		const dirtyImage = MediaStore.getDirtyImage();
+		let numOfImagesToUpdate = 0;
+
 		isVisualEditMode = ! editor.isHidden();
 
 		if ( isVisualEditMode ) {
```
Comment (artpi) [inline client/components/tinymce/plugins/media/plugin.jsx]: I dont know if we care about this case or not, but when I:
1. Edit image
2. Click done
3. Click outside image gallery
4. Click on HTML view

I end up in html view before callback kicks in, Im in html mode now and I have old image in the source.
It's actually easy to do on the slow wifi Im on

File: client/lib/media/store.js [hunk 15]
```diff
@@ -17,7 +17,8 @@ import { isItemBeingUploaded } from 'lib/media/utils';
  */
 const MediaStore = {
 	_media: {},
-	_pointers: {}
+	_pointers: {},
+	_dirtyImageRef: null
```
Comment (artpi) [inline client/lib/media/store.js]: Wouldnt it be better if that were an array?
You can edit multiple images I think

File: client/lib/media/actions.js [hunk 16]
```diff
@@ -260,7 +260,9 @@ MediaActions.update = function( siteId, item, editMediaFile = false ) {
 				type: 'RECEIVE_MEDIA_ITEM',
 				error: error,
 				siteId: siteId,
-				data: data
+				data: editMediaFile
+					? { ...data, isDirty: true }
```
Comment (lamosty) [inline client/lib/media/actions.js]: Do we need it, though? We are only doing things if `isDirty` is `true`. Otherwise, nothing happens.

Comment (artpi) [inline client/lib/media/actions.js]: That's true, it's just much more confusing and right now I have no idea if nothing happens..

**Comment by artpi** [comment]
Nice work @lamosty !
Code is very confusing, but I tested and works pretty nice apart from the html view issue: 
https://github.com/Automattic/wp-calypso/pull/9093#discussion_r86593976

File: client/components/tinymce/plugins/media/plugin.jsx (hunk 5 above)
Comment (gwwar) [inline client/components/tinymce/plugins/media/plugin.jsx]: Oh looks like we can use `editor.wpSetImgCaption`:
```
editor.wpSetImgCaption = function( content ) {
        return parseShortcode( content );
};
```

File: client/components/tinymce/plugins/media/plugin.jsx [hunk 17]
```diff
@@ -279,6 +406,10 @@ function mediaButton( editor ) {
 			// autosave take effect
 			editor.fire( 'change' );
 		}
+
+		if ( numOfImagesToUpdate === 0 ) {
```
Comment (gwwar) [inline client/components/tinymce/plugins/media/plugin.jsx]: Maybe keep a list of dirty media ids, and update this out of the image update loop.

File: client/components/tinymce/plugins/media/plugin.jsx [hunk 18]
```diff
@@ -145,10 +208,19 @@ function mediaButton( editor ) {
...
 
+		const dirtyImage = MediaStore.getDirtyImage();
+		let numOfImagesToUpdate = 0;
+
 		isVisualEditMode = ! editor.isHidden();
 
 		if ( isVisualEditMode ) {
-			images = editor.dom.select( 'img' );
+			if ( dirtyImage ) {
+				images = editor.dom.select( `img.wp-image-${ dirtyImage.ID }` );
```
Comment (gwwar) [inline client/components/tinymce/plugins/media/plugin.jsx]: What happens if I have a queued transient update, plus a dirty image?

Comment (gwwar) [inline client/components/tinymce/plugins/media/plugin.jsx]: Unless there's a huge perf difference, I think it's fine to keep it as it was and iterate over each.

File: client/lib/media/store.js [hunk 19]
```diff
@@ -62,6 +67,14 @@ function receivePage( siteId, items ) {
 	} );
 }
 
+MediaStore.getDirtyImage = () => {
```
Comment (gwwar) [inline client/lib/media/store.js]: Let's avoid keeping a ref of this.

File: client/lib/media/store.js (hunk 15 above)
Comment (lamosty) [inline client/lib/media/store.js]: You can edit only one image at once.

File: client/lib/media/store.js (hunk 19 above)
Comment (lamosty) [inline client/lib/media/store.js]: How do we get it otherwise? By looping through all the images in a post/page and detecting whether it is dirty or not?

File: client/components/tinymce/plugins/media/plugin.jsx (hunk 14 above)
Comment (lamosty) [inline client/components/tinymce/plugins/media/plugin.jsx]: Great catch!

I fixed this partly in 1c7926cea05f25dbf19d47ace4fdc8e683bbd488. Now, image is updated. However, the caption is not resized as it should be. Will look into this.

File: client/components/tinymce/plugins/media/plugin.jsx [hunk 20]
```diff
@@ -190,22 +221,50 @@ function mediaButton( editor ) {
...
+						'ID',
+						'URL',
+						useMediaSize ? 'width' : '',
+						useMediaSize ? 'height' : ''
+					),
+					{
+						transient: !! media.transient
+					}
+				);
 				const options = assign( {}, current.appearance, {
```
Comment (retrofox) [inline client/components/tinymce/plugins/media/plugin.jsx]: can we use `{ ...media, ... }` instead of `assign()`?

**Comment by lamosty** [comment]
The PR is ready for another review.

File: client/components/tinymce/plugins/media/plugin.jsx [hunk 21]
```diff
@@ -44,6 +44,9 @@ import { ModalViews } from 'state/ui/media-modal/constants';
 var REGEXP_IMG = /<img\s[^>]*\/?>/ig,
 	SIZE_ORDER = [ 'thumbnail', 'medium', 'large', 'full' ];
 
+let lastDirtyImage = null,
+	numOfImagesToUpdate = Number.MAX_SAFE_INTEGER;
```
Comment (gwwar) [inline client/components/tinymce/plugins/media/plugin.jsx]: There's no support in IE for MAX_SAFE_INTEGER. Let's set this to something explicit like `null` and check for truthy or 0 before subtracting.

File: client/components/tinymce/plugins/media/plugin.jsx (hunk 11 above)
Comment (gwwar) [inline client/components/tinymce/plugins/media/plugin.jsx]: Let's leave a comment in the code, if it's worth explaning here :D

File: client/components/tinymce/plugins/media/plugin.jsx (hunk 21 above)
Comment (lamosty) [inline client/components/tinymce/plugins/media/plugin.jsx]: Good idea, thanks.

File: client/components/tinymce/plugins/media/plugin.jsx [hunk 22]
```diff
@@ -234,32 +298,43 @@ function mediaButton( editor ) {
...
-				imgString = img.outerHTML;
+			let mediaString = '';
+
+			if ( mediaHasCaption ) {
+				if ( captionNode.outerHTML ) {
+					mediaString = captionNode.outerHTML;
+
+					captionNode.outerHTML = event.content;
+				} else {
+					mediaString = captionNode;
```
Comment (lamosty) [inline client/components/tinymce/plugins/media/plugin.jsx]: Oh, actually this doesn't make sense.

So, initially, the `if` condition was used to differentiate between HTML mode and Visual mode. In Visual mode, we navigate between DOM nodes so we have to use `outerHTML` of a node.

However, in HTML mode, we use `REGEXP_IMG` to capture all the `img` tags so there is no `outerHTML` method on `img`.

In my case, when an image has caption shortcode and we are in HTML mode, the code will replace only the `` tags in the string leaving out the `[caption]` shortcode.

This is a bug @artpi found before. Maybe we could leave it for another PR? Since then, I'll probably remove unneeded code.

Comment (gwwar) [inline client/components/tinymce/plugins/media/plugin.jsx]: I did test this behavior on prod with loading images. (We do handle this properly in the html mode). We can leave this for another PR but I think we'd want this fixed quickly

File: client/components/tinymce/plugins/media/plugin.jsx [hunk 23]
```diff
@@ -234,32 +298,45 @@ function mediaButton( editor ) {
...
 			// instance in the undo stack.
 			//
 			// See: https://github.com/tinymce/tinymce/blob/4.2.4/js/tinymce/classes/EditorUpload.js#L49-L53
 			editor.undoManager.data = editor.undoManager.data.map( function( level ) {
-				level.content = level.content.replace( imgString, event.content );
+				level.content = level.content.replace( mediaString, event.content );
 				return level;
 			} );
+
+			if ( numOfImagesToUpdate ) {
```
Comment (gwwar) [inline client/components/tinymce/plugins/media/plugin.jsx]: We probably won't reach this case, but just a warning that negative numbers are truthy.

Comment (lamosty) [inline client/components/tinymce/plugins/media/plugin.jsx]: Oh, good to know!. However, `0` is not so it'll stop there.

**Comment by gwwar** [comment]
Alright tested this in IE11 too, and I think this is pretty safe to merge. As you mentioned we still have bugs #9183 #9189 after landing, but we can tackle them as a next step.
